### PR TITLE
NMS-10476: Document data types

### PR DESCRIPTION
**Pull request by mmahacek**
### All Contributors

* [X] Have you read and followed our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [X] Have you [made an issue in the OpenNMS issue tracker](https://issues.opennms.org)?<br>If so, you should:
  1. update the title of this PR to be of the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
* [X] Have you made a comment in that issue which points back to this PR?
* [X] Have you updated the JIRA link at the bottom of this comment to link to your issue?
* [X] If this PR targets a `foundation-*` branch, does it avoid changing files in `$OPENNMS_HOME/etc/`?

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-10476

### CODE REVIEW

File: docs/modules/operation/pages/performance-data-collection/data-types.adoc [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+= Data Types
+
+{page-component-title} supports collecting a variety of different data types as defined by https://datatracker.ietf.org/doc/html/rfc2578#section-7.1[RFC 2578].
+When defining data collection definition files, use the closest match to the type of data provided by the node.
+
+== Counter
+
+The counter type represents values that that continue to increase over time.
+This type may be presented from the device with a non-negative integer with a maximum of either a 32-bit or 64-bit upper limit.
+{page-component-title} stores both of these the same, and `COUNTER`, `COUNTER32`, and `COUNTER64` may be used interchangeably in data collection definitions to refer to this data type.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/performance-data-collection/data-types.adoc:11]:
```suggestion
{page-component-title} stores both of these the same, and you can use `COUNTER`, `COUNTER32`, and `COUNTER64` interchangeably in data collection definitions to refer to this data type.
```

File: docs/modules/operation/pages/performance-data-collection/data-types.adoc [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+When a counter reaches the upper limit, the value will reset to zero and continue incrementing.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/performance-data-collection/data-types.adoc:12]:
```suggestion
When a counter reaches the upper limit, the value resets to zero and continues to increment.
```

File: docs/modules/operation/pages/performance-data-collection/data-types.adoc [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+== Gauge
+
+The gauges type represents values that may fluctuate up or down to reflect the current state of a given metric.
+This type may be presented from the device with a non-negative number with a maximum of either a 32-bit or 64-bit upper limit.
+{page-component-title} stores both of these the same, and `GAUGE`, `GAUGE32`, and `GAUGE64` may be used interchangeably in data collection definitions to refer to this data type.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/performance-data-collection/data-types.adoc:18]:
```suggestion
{page-component-title} stores both of these the same, and you can use `GAUGE`, `GAUGE32`, and `GAUGE64` interchangeably in data collection definitions to refer to this data type.
```

File: docs/modules/operation/pages/performance-data-collection/data-types.adoc [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+== Integer
+
+The integers type represents values that represent a number between -2^31 and 2^31-1 (-2147483648 to 2147483647 decimal).
```
Comment (Bonrob2) [inline docs/modules/operation/pages/performance-data-collection/data-types.adoc:22]:
```suggestion
The integer type represents values that represent a number between -2^31 and 2^31-1 (-2147483648 to 2147483647 decimal).
```

File: docs/modules/operation/pages/performance-data-collection/data-types.adoc [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+{page-component-title} stores both of these the same, and `INTEGER`, `INTEGER32`, and `INTEGER64` may be used interchangeably in data collection definitions to refer to this data type.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/performance-data-collection/data-types.adoc:23]:
```suggestion
{page-component-title} stores both of these the same, and you can use `INTEGER`, `INTEGER32`, and `INTEGER64` interchangeably in data collection definitions to refer to this data type.
```

File: docs/modules/operation/pages/performance-data-collection/data-types.adoc [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+Integers may also be used to represent a named-number enumeration status value.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/performance-data-collection/data-types.adoc:24]:
```suggestion
You can also use integers to represent a named-number enumeration status value.
```

File: docs/modules/operation/pages/performance-data-collection/data-types.adoc [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+== IpAddress
+
+The IpAddress type represents a 32-bit internet address.
+
+NOTE: The IpAddress type is a tagged type for historical reasons.
+Network addresses should be represented using a `STRING` value.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/performance-data-collection/data-types.adoc:31]:
```suggestion
Use a `STRING` value to represent network addresses.
```

File: docs/modules/operation/pages/performance-data-collection/data-types.adoc [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+== OctetString
+
+The octetstring type represents arbitrary binary or textual data presented as a hex string.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/performance-data-collection/data-types.adoc:35]:
```suggestion
The OctetString type represents arbitrary binary or textual data presented as a hex string.
```

File: docs/modules/operation/pages/performance-data-collection/data-types.adoc [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+== Opaque
+
+The Opaque type is provided solely for backward-compatibility, and shall not be used for newly-defined object types.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/performance-data-collection/data-types.adoc:39]:
```suggestion
The Opaque type is provided solely for backward-compatibility.
Do not use it for newly-defined object types.
```

File: docs/modules/operation/pages/performance-data-collection/data-types.adoc [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+== String
+
+String values are stored differently from the other numeric data types.
+Only the most recent string value is persisted to disk.
+Stored values can be retrieved using the xref:development:rest/resources.adoc[].
```
Comment (Bonrob2) [inline docs/modules/operation/pages/performance-data-collection/data-types.adoc:45]:
```suggestion
Retrieve stored values with the xref:development:rest/resources.adoc[Resources API].
```

File: docs/modules/operation/pages/performance-data-collection/data-types.adoc [hunk 11]
```diff
@@ -0,0 +1,49 @@
+[[data-types]]
+= Data Types
+
+{page-component-title} supports collecting a variety of different data types as defined by https://datatracker.ietf.org/doc/html/rfc2578#section-7.1[RFC 2578].
+When defining data collection definition files, use the closest match to the type of data provided by the node.
+
+== Counter
+
+The counter type represents values that that continue to increase over time.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/performance-data-collection/data-types.adoc:9]:
```suggestion
The counter type represents values that continue to increase over time.
```

File: docs/modules/operation/pages/performance-data-collection/data-types.adoc [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+This type may be presented from the device with a non-negative integer with a maximum of either a 32-bit or 64-bit upper limit.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/performance-data-collection/data-types.adoc:10]:
```suggestion
The device may present this type with a non-negative integer with a maximum of either a 32-bit or 64-bit upper limit.
```

File: docs/modules/operation/pages/performance-data-collection/data-types.adoc [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+{page-component-title} stores both of these the same, and `COUNTER`, `COUNTER32`, and `COUNTER64` may be used interchangeably in data collection definitions to refer to this data type.
+When a counter reaches the upper limit, the value will reset to zero and continue incrementing.
+
+== Gauge
+
+The gauges type represents values that may fluctuate up or down to reflect the current state of a given metric.
+This type may be presented from the device with a non-negative number with a maximum of either a 32-bit or 64-bit upper limit.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/performance-data-collection/data-types.adoc:17]:
```suggestion
The device may present this type with a non-negative number with a maximum of either a 32-bit or 64-bit upper limit.
```

**Review by Bonrob2 — CHANGES_REQUESTED** [review]

**Comment by mmahacek** [comment]
@Bonrob2 I have this being added to the Operation section for collection.  Should it be moved into the Reference section instead?

**Comment by mmahacek** [comment]
@Bonrob2 I have this being added to the Operation section for collection.  Should it be moved into the Reference section instead?

**Comment by Bonrob2** [comment]
> @Bonrob2 I have this being added to the Operation section for collection. Should it be moved into the Reference section instead?

It makes sense in the reference section, but we would want to keep a stub in the operations section where we could put an example, if appropriate, then cross reference to the reference section.

File: docs/modules/operation/pages/performance-data-collection/data-types.adoc [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+Use a `STRING` value to represent network addresses.
+
+== OctetString
+
+The OctetString type represents arbitrary binary or textual data presented as a hex string.
+
+== Opaque
+
+The Opaque type is provided solely for backward-compatibility.
+Do not use it for newly-defined object types.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/performance-data-collection/data-types.adoc:40]:
```suggestion
Do not use it for newly defined object types.
```

**Review by Bonrob2 — CHANGES_REQUESTED** [review]
Just one change. Otherwise, looks good.

**Pull request merged**